Create a New Template - Documentation (#364)
microsoft/cobalt: docs/3_NEW_TEMPLATE.md

# 3. Cobalt Templating From Scratch

## 3.1 Overview

> Per the [quickstart guide](./2_QUICK_START_GUIDE.md), we demonstrated that you can get pretty far with *Cobalt* by simply relying on any out of the box *Cobalt Infrastructure Template* or *CIT* (/kɪt/). However, for several reasons, you may very well have unique infrastructure scenarios that require you to extract more use out of *Cobalt*. Therefore, we recommend building a *Cobalt Infrastructure Template* from scratch in order to cultivate a deeper understanding of what *Cobalt* has to offer.

Reviewer: Here are some suggestions for making this a little less wordy. I like the overall intent a lot!
```suggestion
Per the [quickstart guide](./2_QUICK_START_GUIDE.md), we demonstrated how easy it is to deploy an existing *Cobalt Infrastructure Template* or *CIT* (/kɪt/). However, it is likely that you will need to develop your own *CIT* by composing the foundational infrastructure modules that ship with *Cobalt*. This guide will walk you through building a *Cobalt Infrastructure Template* from scratch in order to cultivate a deeper understanding of what *Cobalt* has to offer.
```

> A major core feature of *Cobalt* is that it offers a library of Terraform based *Cobalt Module*s that allow you to create and build-up *CIT*s. The act of creating a *CIT* from scratch will always involve thoughtfully choosing a mix of *Cobalt Module*s that already exist or were created by you. This section of the walkthrough will be an exercise in building *CIT*s for Cobalt. Happy templating! 😄

Reviewer: ```suggestion
A core feature of *Cobalt* is that it offers a library of Terraform based *Cobalt Module*s that allow you to create and build-up *CIT*s by composing *Modules*. Cobalt ships with a number of *Modules*, but you can also write your own. The act of creating a *CIT* from scratch will always involve thoughtfully choosing a mix of *Module*s that already exist or were created by you. This section of the walkthrough will be an exercise in building *CIT*s for Cobalt. Happy templating! 😄
```

| Cobalt Infrastructure Template | CIT Instantiating an [Azure service-plan](./../infra/modules/providers/azure/service-plan/README.md) Module |
|----------|----------|
|[az-hello-world](./../infra/templates/az-hello-world/README.md)| ![image](https://user-images.githubusercontent.com/10041279/67301762-82123500-f4b5-11e9-9bff-8dc07a4fe001.png) |
|[az-service-single-region](./../infra/templates/az-hello-world/README.md)| ![image](https://user-images.githubusercontent.com/10041279/67302608-bf2af700-f4b6-11e9-9add-846bd2df42be.png) |
|[az-isolated-service-single-region](./../infra/templates/az-hello-world/README.md)| ![image](https://user-images.githubusercontent.com/10041279/67302203-2ac09480-f4b6-11e9-839f-19d40abd51ae.png) |

> The above table demonstrates in a clear way how *Cobalt Infrastructure Templates* reap the natural benefits of the reusability that's offered by *Cobalt Modules*. This is possible because Terraform modules grant *Cobalt modules* reusability as a feature. Let's experience what it's like to create your own *CIT* from scratch by following the below steps:

Reviewer: IMO, the above list is really nice and should stay within the document, but does not do what is stated in this paragraph.

The benefit of the module, in my mind, is that it distills the otherwise complicated task of properly configuring a set of related resources in a way that is re-usable for many use cases. An example of this could be "configuring a storage container and granting RBAC access to a service principal," or "configuring App Service with a staging slot and also configuring easy auth"

This context is missing, and probably should live higher in the document when you define what a module is and why you might want one when opening up your "box of cobalt"

### **Step 1:** Model your planned infrastructure

> For demonstration purposes, we have already modeled the infrastructure. You will build a *CIT* and title it **az-hello-world-from-scratch** within your codebase. This CIT when ran will create and deploy the Azure resources listed in the description of the below table:

Reviewer: This name is similar to an existing CIT but does something much different -- perhaps `az-functions-hello-world`?

| New CIT Name | Description | Deployment Goal |
|----------|----------|----------|
| **az-hello-world-from-scratch** | A Cobalt Infrastructue Template that when ran creates a basic [Azure Function App](https://docs.microsoft.com/en-us/azure/azure-functions/functions-overview) within an [App Service Plan](https://docs.microsoft.com/en-us/azure/app-service/overview-hosting-plans) accompanied with [Azure Storage](https://azure.microsoft.com/en-us/services/storage/blobs/). | <image src="https://user-images.githubusercontent.com/10041279/67136958-a9d27600-f1f3-11e9-896c-d18f3a287de5.png" width="500" height="200"/> |

> ### **Step 2:** Plan your modular strategy

Reviewer: ```suggestion
### **Step 2:** Plan your module strategy
```

Author: Done!

Once you've modeled your planned infrastructure resources, we recommend answering the following questions:

1. > QUESTION: **"Which portion of my planned infrastructure do I want to roll-up into a reusable module?"**

>     **ANSWER:** Currently, most Cobalt Modules are scoped to a very specific cloud infrastructure resource. In other words, modules encapsulate a configuration that targets a providers very specific cloud service product. The Azure Function Module we are creating will follow this pattern.

Reviewer: ```suggestion
    **ANSWER:** Currently, most Cobalt Modules are scoped to a small group of closely related resources. In other words, modules encapsulate a configuration that targets a providers very specific set of cloud service products. The Azure Function Module we are creating will follow this pattern.
```

Author: Rewording this was a great idea.

1. > QUESTION: **"Does Cobalt have existing reusuable modules configured for any portion of my planned infrastructure?"**

    **ANSWER:** At the time of this walkthrough, there are reusable modules for an Azure App Service Plan and Azure Storage, so you will use those to build part of your *CIT*. However, there is no current *Cobalt Module* configured for deploying an Azure Function. So, yes, a new Cobalt Module needs to be created. Let's design one!

### **Step 3:** Design Your Terraform Based *Cobalt Module*s

> The first step of designing a *Cobalt Module* involves defining a Terraform module's 3 primary elements: input variables, output variables and resources. This will be all done via Terraform's [HCL language](https://learn.hashicorp.com/terraform), a language that grants developers the ability to target multiple cloud providers. Documentation for the HCL language is partitioned by cloud provider. You will become very familiar with Terraform's cloud provider documentation as you learn to use and build your own modules and CITs.

Reviewer: ```suggestion
The first step of designing a *Cobalt Module* involves defining a Terraform module's 3 primary elements: input variables, output variables and resources. This will be all done via Terraform's [HCL language](https://learn.hashicorp.com/terraform), a language that enables developers the ability to declaratively describe infrastructure resources that may exist across multiple cloud providers. Documentation for the HCL language is [located here](https://www.terraform.io/docs/configuration/syntax.html) but the implementation of resources are partitioned by cloud provider. You will become very familiar with Terraform's cloud provider documentation as you learn to use and build your own modules and CITs.
```

1. **Visit the below link**. Use the documentation at the link as a reference for insight into how we are making the below design decisions for the Azure Function *Cobalt Module* in this walkthrough.

    * Terraform - [Azure ARM Provider - Azure Function](https://www.terraform.io/docs/providers/azurerm/r/function_app.html#example-usage-in-a-consumption-plan-)

1. **Define your resources** - Defined below are the resource blocks that will be implemented:

>     | Resource | Description |

Reviewer: This may be nice with another column that describes which module each will live in. it makes the description less wordy, and goes along with the "module planning" strategy you are discussing here

Author: Definitely added a "module" column. Good thinking.

    |--------|-------------|
    | azurerm_function_app | According to the Terraform docs, this is the only resource unique to an Azure Function. This resource block will be declared within the module. |
    | azurerm_app_service_plan | The azure function app needs to live within an app service plan so we have to instantiate the existing module that encapsulates the app service plan. |
    | azurerm_storage_account | The azure function app's ephemeral state needs a dedicated storage account, however, after further inspection, the existing module for the storage account will not be usuable as it does not support a connection string as an output due to security best practices. We'll have to declare the azurerm_storage_account resource directly from the template and avoid using it's module.  |
    | azurerm_resource_group | Most Azure infrastructure lives in a resource group container so this will be a part of the configuration as well.  |

> 1. **Define inputs** - When a CIT instantiates a module, it will configure that module using it's exposed input variable names. These variables will pass values to the attributes of the resource blocks internal to the module. These inputs have also been defined for you below:

Reviewer: ```suggestion
1. **Define inputs** - *CIT*s configure modules using the module input variables. The module will use these to configure the resources it depends on. These inputs have also been defined for you below:
```

Author: Used the suggestion to make this more brief. Thanks for this.

>     | *azurerm_function_app* resource attribute | Scope | Required | Input Variable Name | Description |

Reviewer: This table is a little confusing as it mixes implementation details of the module with the inputs that you are describing.

Would be more clear with 3 columns: (1) Variable name, (2) Description, (3) Default Value

Author: Columns are now (1) input var (2) type (3) default value (4) provider (5) description.

    | name | Input | yes | `azure_function_name_prefix` | A prefix name for appending unique values to the azure function name. |
    | resource_group_name | Input | yes | `resource_group` | Most Azure infrastructure lives in a resource group container of your choice. By making this an input, each module instance can have a different resouce group. |
    | location | Input | yes | `resource_group_location` | The geo-location here should derive from the geo-location that the resource group name lives in. |
    | app_service_plan_id  | Input | yes | `app_service_plan_id` | This input implies that the azure function resource will live within an app service plan. |
    | storage_connection_string | Input | yes | `storage_connection_string` | This is the storage account in which the ephemeral state for an Azure Function will be orchestrated when the endpoint is invoked. |
    | app_settings | Internal | no | `-` | { environment = "hw-from-scratch" } - We will provide a hard-coded key-value pair as an example that does not require an input. Value will not be passed from a CIT. |

> 1. **Define outputs** - A module instance will only output values that it's been pre-configured to output. It's **best practice** to configure module instance outputs so that you can validate expected results. These results are visible in standard out if passed to the template when running the terraform plan and apply steps. These outputs are defined for you below:

Reviewer: ```suggestion
1. **Define outputs** - A module instance will only output values that it's been pre-configured to output. It's **best practice** to configure module outputs because it enables module composition. These results are visible in standard out if passed to the template when running the terraform plan and apply steps. These outputs are defined for you below:
```

>     | *azurerm_function_app* attribute | Scope | Required | Output Variable Name | Description |

Reviewer: Similar comment - more clear with 2 columns: (1) Output Name, (2) Description

    output "azure_function_id" {
        description = "The URLs of the app services created."
        value       = azurerm_function_app.walkthrough.id
    }
    output "azure_function_url" {
        description = "The resource ids of the app services created."
        value       = azurerm_function_app.walkthrough.default_hostname
    }
    output "app_service_type" {
        description = "The type of app service created."
        value       = azurerm_function_app.walkthrough.kind
    }
    ```

### **Step 5:** Implement Azure Hello World From Scratch CIT

Let's implement the Azure Hello World From Scratch CIT by instantiating our new Azure Function Cobalt Module along with the modules that it depends on.

1. Navigate to the infra templates directory (i.e. ./infra/templates) and execute the following commands to wire up your new CIT:

    ```bash
    # Create a directory called "function-app"
    mkdir -p ./az-hello-world-from-scratch
    # Navigate to that directory
    cd az-hello-world-from-scratch
    # Copy generic files
    cp ./../az-hello-world/backend.tf backend.tf

>     cp ./../az-hello-world/versions.tf versions.tf

Reviewer: Why do we need this?

    # Create a commons.tf, main.tf, variables.tf, outputs.tf and terraform.tfvars

>     touch variables.tf touch commons.tf touch main.tf touch outputs.tf touch terraform.tfvars

Reviewer: ```suggestion
    touch variables.tf && touch commons.tf && touch main.tf && touch outputs.tf && touch terraform.tfvars
```

        type = string
        default = "Y1"
    }

    variable "service_plan_kind" {
        description = "The kind of Service Plan to be created. i.e. FunctionApp"
        type = string
        default = "FunctionApp"
    }
    ```

1. Open the commons.tf file and paste the the following:

    ```HCL
    module "provider" {
        source = "../../modules/providers/azure/provider"
    }

    resource "random_string" "workspace_scope" {
        keepers = {
            # Generate a new id each time we switch to a new workspace or app id
            ws_name = replace(trimspace(lower(terraform.workspace)), "_", "-")
            app_id  = replace(trimspace(lower(var.name)), "_", "-")
        }

        length  = max(1, var.randomization_level) // error for zero-length
        special = false
        upper   = false
    }

    locals {
        // sanitize names
        app_id  = random_string.workspace_scope.keepers.app_id
        region  = replace(trimspace(lower(var.resource_group_location)), "_", "-")
        ws_name = random_string.workspace_scope.keepers.ws_name
        suffix  = var.randomization_level > 0 ? "-${random_string.workspace_scope.result}" : ""

>         // base name for resources, name constraints documented here: https://docs.microsoft.com/en-us/azure/architecture/best-practices/naming-conventions

Reviewer: There is a lot of context regarding randomizing resource names that seems strange *unless you have been bitten by it in the future*. It may be worthwhile to add context about why we do it

Author: Excellent point! Added brief comments about in Step 5 of step 5!
